### DISCUSSION TITLE
problems with generic `TEvent extends ...`

### DISCUSSION THREAD

**Discussion by mikaelkaron** [statelyai/xstate#3556] [Q&A] [answered]
I'm trying to write a function that takes a `machine` that supports at _least_ a set of defined events. The problem I'm having is when I try to further generalize the `TEvent` parameter using `extends`:

```typescript
import { interpret, StateMachine, StateSchema } from "xstate";

export type TestEventObject = { type: "EVENT1" };

export const test1 = <
  TContext,
  TSchema extends StateSchema<TContext>,
  TEvent extends TestEventObject
>(
  machine: StateMachine<TContext, TSchema, TEvent>
) => {
  const service = interpret(machine);

  service.send({
    type: "EVENT1"
  });
};

export const test2 = <
  TContext,
  TSchema extends StateSchema<TContext>
>(
  machine: StateMachine<TContext, TSchema, TestEventObject>
) => {
  const service = interpret(machine);

  service.send({
    type: "EVENT1"
  });
};
```

In the above example `test2` compiles fine while `test1` gets this error:

```
Argument of type '{ type: "EVENT1"; }' is not assignable to parameter of type 'SingleOrArray<Event<TEvent>> | Event<TEvent>'.
  Object literal may only specify known properties, and 'type' does not exist in type 'never'.ts(2345)
```

I'm guessing this is just a common typescript problem and not really an xstate problem, but it would be nice to know how to solve this.

**Comment by with-heart** [comment, marked as answer]
The type constraint should go on the `machine` argument, not in the generic types:

```ts
export const test1 = <
  TContext,
  TSchema extends StateSchema<TContext>,
  TEvent extends EventObject
>(
  machine: StateMachine<TContext, TSchema, TEvent | TestEventObject>,
) => {
  const service = interpret(machine)

  service.send({
    type: "EVENT1",
  })
}
```

The generic types are what allows typescript to _infer_ `TContext`, `TSchema`, and `TEvent` types from the `StateMachine` type of the `machine` argument:

```ts
const machine = createMachine<SomeContext, SomeSchema, SomeEvent>({})
//    ^ StateMachine<SomeContext, SomeSchema, SomeEvent>

// the following are equivalent:

// manually pass the types…
test<SomeContext, SomeSchema, SomeEvent>(machine)

// …or let typescript infer them from `StateMachine`
test(machine)
```

Creating a union of `TEvent | TestEventObject` in the `machine` argument means that the provided `StateMachine` must include `TestEventObject` in its events.

To put it another way, you're constraining the type of `StateMachine`, not the type of `test1`:

```ts
// good to go!
const machine: StateMachine<any, any, TestEventObject> = createMachine({
  schema: { events: {} as TestEventObject }
})

// type error
const machine: StateMachine<any, any, TestEventObject> = createMachine({
  schema: { events: {} as SomeOtherEvent }
})
```

PS: Here's [a CodeSandbox reproduction](https://codesandbox.io/s/eloquent-burnell-zbk0up?file=/src/index.tsx)

**Reply by mikaelkaron** [reply]
That is an _awesome_ answer, thank you so much!

It makes sense (now that I look at your solution) that it should be that way. Now to level up on some generics (maybe I should buy some @mattpocock courses)

**Reply by mikaelkaron** [reply]
Ok, so there's another _small_ problem. If I add extra properties to `TestEventObject` I can't get them working in the `subscribe`.

```typescript
export const test = <
  TContext,
  TSchema extends StateSchema<TContext>,
  TEvent extends EventObject
>(
  machine: StateMachine<TContext, TSchema, TEvent | TestEventObject>
) => {
  const service = interpret(machine);

  service.subscribe((state) => {
    const { event } = state;
    if (event.type === "EVENT1") {
      console.log(event.data);
    }
  });

  service.send({
    type: "EVENT1",
    data: {}
  });
};
```

I [forked your sandbox](https://codesandbox.io/s/competent-robinson-pb2ne8?file=/src/index.tsx) to illustrate

**Reply by Andarist** [reply]
This is a TS limitation~. We can't express that your `TEvent` **won't** contain an event with `type: "EVENT1"`. This is still a valid call:
```ts
const myMachine = createMachine({
  schema: {
    events: {} as
      | TestEventObject
      | SomeOtherEvent
      | { type: "EVENT1"; noDataProperty: true }
  }
});

test(myMachine);
```
So within the generic context of the `test`'s implementation TypeScript can't "prove" that you will have a `.data` property there. The `event.type` is just a string because it's a common type for this property of all union members (and that includes the constraint of `TEvent` which is just `{ type: string }`), so in this case, it won't be able to use this to narrow this union.

In situations like this the only workaround is to use casting within the generic implementation.

**Reply by mikaelkaron** [reply]
> We can't express that your `TEvent` won't contain an event with `type: "EVENT1"`

Interesting. So theoretically if you could _exclude_ all the `type: ...` clashes between `TEvent` and `TestEventObject` it would work as expected?

**Reply by Andarist** [reply]
Theoretically - yes.
